pkg/controller/servicebinding/secret.go

```diff
@@ -10,98 +12,108 @@ import (
+	ownerReference metav1.OwnerReference,
+) (*unstructured.Unstructured, error) {
+	secretHash, err := name(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	secret := &corev1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
-			Namespace:       s.ns,
-			Name:            s.name,
+			Namespace:       ns,
+			Name:            prefix + "-" + secretHash,
 			OwnerReferences: []metav1.OwnerReference{ownerReference},
 		},
 		Data: payload,
 	}
 
 	gvk := corev1.SchemeGroupVersion.WithKind(secretKind)
-	u, err := converter.ToUnstructuredAsGVK(secretObj, gvk)
+	u, err := converter.ToUnstructuredAsGVK(secret, gvk)
 	if err != nil {
 		return nil, err
 	}
 
-	resourceClient := s.buildResourceClient()
-
-	logger.Debug("Attempt to create secret...")
-	existingSecret, err := s.get()
+	resourceClient := buildResourceClient(client, ns)
+	expectedSecret, err := read(resourceClient, secret.Name, ns)
 	if err != nil {
 		if errors.IsNotFound(err) {
-			_, err := resourceClient.Create(u, metav1.CreateOptions{})
+			secretNew, err := resourceClient.Create(u, metav1.CreateOptions{})
 			if err != nil {
-				logger.Error(err, "Error creating secret")
 				return nil, err
 			}
-			logger.Info("Secret created")
-			return u, nil
+			return secretNew, nil
 		}
 		return nil, err
 	}
-	existingSecretData, _, _ := unstructured.NestedMap(existingSecret.Object, "data")
-	payloadInterim := make(map[string]interface{})
-	for k, v := range payload {
-		payloadInterim[k] = base64.StdEncoding.EncodeToString(v)
+
+	expectedSecretData, _, err := unstructured.NestedMap(expectedSecret.Object, "data")
+	if err != nil {
+		return nil, err
+	}
+	expectedData := make(map[string][]byte)
+	for k, v := range expectedSecretData {
+		expectedData[k] = []byte(v.(string))
+	}
+	expectedDataHash, err := name(expectedData)
+	if err != nil {
+		return nil, err
 	}
-	comparisonResult := nestedMapComparison(existingSecretData, payloadInterim)
-	if comparisonResult.Success {
-		logger.Debug("Secret data is same. Skip Update")
-	} else {
-		logger.Info("Secret data is different; update secret", "Diff", comparisonResult.Diff)
-		_, err = resourceClient.Update(u, metav1.UpdateOptions{})
+
+	// compare current and existing secret name and data hash
+	if expectedSecret.GetName() != secret.Name || expectedDataHash != secretHash {
+		secretNew, err := resourceClient.Update(u, metav1.UpdateOptions{})
 		if err != nil {
 			return nil, err
 		}
+		return secretNew, nil
 	}
+
 	return u, nil
 }
 
-// get an unstructured object from the secret handled by this component. It can return errors in case
+// read an unstructured object from the secret handled by this component. It can return errors in case
 // the API server does.
-func (s *secret) get() (*unstructured.Unstructured, error) {
-	resourceClient := s.buildResourceClient()
-	u, err := resourceClient.Get(s.name, metav1.GetOptions{})
+func read(resourceClient dynamic.ResourceInterface, name, namespace string) (*unstructured.Unstructured, error) {
```

Comment:
Before this symbol was scoped by the `secret` type, but now it doesn't and the function (which is available to the entire `servicebinding` package) has a very opaque name. Please change it to `getSecret(resourceClient, name, namespace)` or something more meaningful instead.